go/pools/smartconnpool/pool.go

```diff
@@ -595,7 +596,9 @@ func (pool *ConnPool[C]) get(ctx context.Context) (*Pooled[C], error) {
 	// to other clients, wait until one of the connections is returned
 	if conn == nil {
 		start := time.Now()
-		conn, err = pool.wait.waitForConn(ctx, nil)
+		conn, err = pool.wait.waitForConn(ctx, nil, func() bool {
+			return pool.close == nil || pool.capacity.Load() == 0
```

Comment:
if `pool.close` safe to access?

Reply:
I believe it's safe to access. Would be even safer to add a nil check on `pool` though :) 

Yeah, I think we should add a method for it.

Reply:
I changed this in https://github.com/vitessio/vitess/pull/18713/commits/7871067d5f35da1e0f546ce20f85cd339203a13e, and further tweaked it in some of the followup PRs.